File: kdroidext/src/main/java/com/nowfal/kdroidext/pref/KPrefHelper.java
package com.nowfal.kdroidext.pref;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.SharedPreferences;
import android.os.Build;
import android.preference.PreferenceManager;

import java.util.Map;
import java.util.Set;

/**
 * Created by Viral Patel on 01-Feb-16.
 */
public class KPrefHelper {

    private final SharedPreferences settings;

    //region Default Values for various types
    private int intDefaultVal = 0;
    private long longDefaultVal = 0;
    private float floatDefaultVal = 0;
    private boolean boolDefaultVal = false;
    private String stringDefaultVal = "";
    private Set<String> stringSetDefaultVal = null;
    //endregion

    //region Getters and Setters for default Values
    public int getIntDefaultVal() {
        return intDefaultVal;
    }

    public void setIntDefaultVal(int intDefaultVal) {
        this.intDefaultVal = intDefaultVal;
    }

    public long getLongDefaultVal() {
        return longDefaultVal;
    }

    public void setLongDefaultVal(long longDefaultVal) {
        this.longDefaultVal = longDefaultVal;
    }

    public float getFloatDefaultVal() {
        return floatDefaultVal;
    }

    public void setFloatDefaultVal(float floatDefaultVal) {
        this.floatDefaultVal = floatDefaultVal;
    }

    public boolean isBoolDefaultVal() {
        return boolDefaultVal;
    }

    public void setBoolDefaultVal(boolean boolDefaultVal) {
        this.boolDefaultVal = boolDefaultVal;
    }

    public String getStringDefaultVal() {
        return stringDefaultVal;
    }

    public void setStringDefaultVal(String stringDefaultVal) {
        this.stringDefaultVal = stringDefaultVal;
    }

    public Set<String> getStringSetDefaultVal() {
        return stringSetDefaultVal;
    }

    public void setStringSetDefaultVal(Set<String> stringSetDefaultVal) {
        this.stringSetDefaultVal = stringSetDefaultVal;
    }
    //endregion

    //region Constructors
    /**
     * Uses the default shared preferences
     */
    public KPrefHelper(Context context) {
        settings = PreferenceManager.getDefaultSharedPreferences(context);
    }

    /**
     * creates a new preference file with specified name
     * @param context
     * @param preferenceFileName
     */
    public KPrefHelper(Context context, String preferenceFileName) {
        this(context, preferenceFileName, 0);
    }

    /**
     * creates a new preference file with specified name and mode
     * @param context
     * @param preferenceFileName
     * @param mode
     */
    public KPrefHelper(Context context, String preferenceFileName, int mode) {
        settings = context.getSharedPreferences(preferenceFileName, mode);
    }
    //endregion

    //#TODO Builder Pattern here can be helpful in making the use of this much elegant
    //region Put Methods
    public void putInt(String key, int val) {
        settings.edit().putInt(key, val).apply();
    }

    public void putString(String key, String val) {
        settings.edit().putString(key, val).apply();
    }

    public void putBoolean(String key, boolean val) {
        settings.edit().putBoolean(key, val).apply();
    }

    public void putFloat(String key, float val) {
        settings.edit().putFloat(key, val).apply();
    }

    public void putLong(String key, long val) {
        settings.edit().putLong(key, val).apply();
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public void putStringSet(String key, Set<String> val) {
        settings.edit().putStringSet(key, val).commit();
    }
    //endregion

    //region Get Methods
    public Map<String, ?> getAll() {
        return settings.getAll();
    }

    public int getInt(String key, int defaultValue) {
        return settings.getInt(key, defaultValue);
    }

    public int getInt(String key) {
        return settings.getInt(key, intDefaultVal);
    }

    public String getString(String key, String defaultValue) {
        return settings.getString(key, defaultValue);
    }

    public String getString(String key) {
        return settings.getString(key, stringDefaultVal);
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        return settings.getBoolean(key, defaultValue);
    }

    public boolean getBoolean(String key) {
        return settings.getBoolean(key, boolDefaultVal);
    }

    public float getFloat(String key, float defaultValue) {
        return settings.getFloat(key, defaultValue);
    }

    public float getFloat(String key) {
        return settings.getFloat(key, floatDefaultVal);
    }

    public long getLong(String key, long defaultValue) {
        return settings.getLong(key, defaultValue);
    }

    public long getLong(String key) {
        return settings.getLong(key, longDefaultVal);
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public Set<String> getStringSet(String key, Set<String> defaultValue) {
        return settings.getStringSet(key, defaultValue);
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    public Set<String> getStringSet(String key) {
        return settings.getStringSet(key, stringSetDefaultVal);
    }
    //endregion

    //region Listener registering and unregistering methods
    public void registerListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        settings.registerOnSharedPreferenceChangeListener(listener);
    }

    public void unregisterListener(SharedPreferences.OnSharedPreferenceChangeListener listener) {
        settings.unregisterOnSharedPreferenceChangeListener(listener);
    }
    //endregion
}

File: kdroidext/src/main/java/in/nowfal/kdroidext/pref/KPrefHelper.kt
package `in`.nowfal.kdroidext.pref

import android.annotation.TargetApi
import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import android.preference.PreferenceManager

/**
 * Created by Viral Patel on 01-Feb-16.
 */
class KPrefHelper {

    private val settings: SharedPreferences

    //region Default Values for various types
    //endregion

    //region Getters and Setters for default Values
    var intDefaultVal = 0
    var longDefaultVal: Long = 0
    var floatDefaultVal = 0f
    var isBoolDefaultVal = false
    var stringDefaultVal = ""
    var stringSetDefaultVal: Set<String>? = null
    //endregion

    //region Get Methods
    val all: Map<String, *>
        get() = settings.all
    //endregion

    //region Constructors
    /**
     * Uses the default shared preferences
     */
    constructor(context: Context) {
        settings = PreferenceManager.getDefaultSharedPreferences(context)
    }

    /**
     * creates a new preference file with specified name and mode
     * @param context
     * @param preferenceFileName
     * @param mode
     */
    @JvmOverloads
    constructor(context: Context, preferenceFileName: String, mode: Int = 0) {
        settings = context.getSharedPreferences(preferenceFileName, mode)
    }
    //endregion

    //#TODO Builder Pattern here can be helpful in making the use of this much elegant
    //region Put Methods
    fun putInt(key: String, `val`: Int) {
        settings.edit().putInt(key, `val`).apply()
    }

    fun putString(key: String, `val`: String) {
        settings.edit().putString(key, `val`).apply()
    }

    fun putBoolean(key: String, `val`: Boolean) {
        settings.edit().putBoolean(key, `val`).apply()
    }

    fun putFloat(key: String, `val`: Float) {
        settings.edit().putFloat(key, `val`).apply()
    }

    fun putLong(key: String, `val`: Long) {
        settings.edit().putLong(key, `val`).apply()
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    fun putStringSet(key: String, `val`: Set<String>) {
        settings.edit().putStringSet(key, `val`).apply()
    }

    fun getInt(key: String, defaultValue: Int): Int {
        return settings.getInt(key, defaultValue)
    }

    fun getInt(key: String): Int {
        return settings.getInt(key, intDefaultVal)
    }

    fun getString(key: String, defaultValue: String): String? {
        return settings.getString(key, defaultValue)
    }

    fun getString(key: String): String? {
        return settings.getString(key, stringDefaultVal)
    }

    fun getBoolean(key: String, defaultValue: Boolean): Boolean {
        return settings.getBoolean(key, defaultValue)
    }

    fun getBoolean(key: String): Boolean {
        return settings.getBoolean(key, isBoolDefaultVal)
    }

    fun getFloat(key: String, defaultValue: Float): Float {
        return settings.getFloat(key, defaultValue)
    }

    fun getFloat(key: String): Float {
        return settings.getFloat(key, floatDefaultVal)
    }

    fun getLong(key: String, defaultValue: Long): Long {
        return settings.getLong(key, defaultValue)
    }

    fun getLong(key: String): Long {
        return settings.getLong(key, longDefaultVal)
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    fun getStringSet(key: String, defaultValue: Set<String>): Set<String>? {
        return settings.getStringSet(key, defaultValue)
    }

    @TargetApi(Build.VERSION_CODES.HONEYCOMB)
    fun getStringSet(key: String): Set<String>? {
        return settings.getStringSet(key, stringSetDefaultVal)
    }
    //endregion

    //region Listener registering and unregistering methods
    fun registerListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        settings.registerOnSharedPreferenceChangeListener(listener)
    }

    fun unregisterListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        settings.unregisterOnSharedPreferenceChangeListener(listener)
    }
    //endregion
}
